File: students/kongshiwei/Kmeans.m
[b] = xlsread('mix.xlsx',1,'A1：D1628');
x = b(:,1);
y = b(:,2);
%c = b(:,3);
%已经将数据集随机排序了所以直接取前23个作为样本
data = [x(1:23,1),y(1:23,1)];
%用于记录点到样本的距离
dist = zeros(1,23);
for k = 1:1000
%用来记录点被分到那个类中
c = zeros(1628,1);
sum = zeros(23,3);
for i = 1:1628
    for j = 1:23
        dist(1,j) = sqrt((x(i,1)-data(j,1))^2+(y(i,1)-data(j,2))^2);
    end
    [mi,index]=min(dist);
    c(i,1) = index;
    sum(c(i,1),1)=x(i)+sum(c(i,1),1);
    sum(c(i,1),2)=y(i)+sum(c(i,1),2);
    sum(c(i,1),3)=sum(c(i,1),3)+1;
end
%重新计算均值
for m = 1:23
    data(m,1) = sum(m,1)/sum(m,3);
    data(m,2) = sum(m,2)/sum(m,3);
end
end
%画图
for i = 1:1628
    rand('seed',c(i,1));
    color = rand(1,3);
    plot(x(i,1),y(i,1),'*','color',color);
    hold on;
end    

File: students/mengjundu/exp1/kmeans.py
#coding=utf-8
from numpy import *
import numpy as np
import random
import matplotlib.pyplot as plt

def loadDataSet(fileName):
    dataMat = []
    data = np.loadtxt(fileName, delimiter=',')
    position = data[:, 0:2]
    dataMat = np.mat(position)
    return dataMat


def distEclud(vecA, vecB):
    return sqrt(sum(power(vecA - vecB, 2)))


def randCent(dataSet, k):
    n = np.shape(dataSet)[1]
    centers = np.mat(np.zeros((k, n)))

    for col in range(n):
        mincol = min(dataSet[:, col])
        maxcol = max(dataSet[:, col])
        centers[:, col] = np.mat(mincol + float(maxcol - mincol) * np.random.rand(k, 1))

    return centers


def kMeans(dataSet, k, distMeas=distEclud, createCent=randCent):
    m = shape(dataSet)[0]
    clusterAssment = mat(zeros((m, 2)))
    centroids = createCent(dataSet, k)
    clusterChanged = True
    while clusterChanged:
        clusterChanged = False
        for i in range(m):
            minDist = inf
            minIndex = -1
            for j in range(k):
                distJI = distMeas(centroids[j, :], dataSet[i, :])
                if distJI < minDist:
                    minDist = distJI;
                    minIndex = j
            if clusterAssment[i, 0] != minIndex:
                clusterChanged = True
            clusterAssment[i, :] = minIndex, minDist ** 2
        print
        centroids
        for cent in range(k):
            ptsInClust = dataSet[nonzero(clusterAssment[:, 0].A == cent)[0]]
            centroids[cent, :] = mean(ptsInClust, axis=0)
    return centroids, clusterAssment


def show(dataSet, k, centroids, clusterAssment):
    from matplotlib import pyplot as plt
    numSamples, dim = dataSet.shape
    mark = ['or', 'ob', 'og', 'ok', '^r', '+r', 'sr', 'dr', '<r', 'pr']
    for i in range(numSamples):
        markIndex = int(clusterAssment[i, 0])
        plt.plot(dataSet[i, 0], dataSet[i, 1], mark[markIndex])
    mark = ['Dr', 'Db', 'Dg', 'Dk', '^b', '+b', 'sb', 'db', '<b', 'pb']
    for i in range(k):
        plt.plot(centroids[i, 0], centroids[i, 1], mark[i], markersize=12)
    plt.show()


def main():
    dataMat = mat(loadDataSet('F:\Course_PR_17-master\experiment1\data\synthetic_data\Aggregation.txt'))
    myCentroids, clustAssing = kMeans(dataMat, 7)
    print
    myCentroids
    show(dataMat, 7, myCentroids, clustAssing)


if __name__ == '__main__':
    main()